### PR TITLE
Use average NTG values for merged cells due to MINPV algorithm

### PR DESCRIPTION
**Pull request by totto82**
The volume weighted averaged NTG values are used to get more reasonable x
and y transmissibilities for the cells effected by the MINPV grid
processing algorithm.

This commit starts the process of supporting the MINPVF keyword
where upscaled properties are used for the merged/filled cells.

The remaining work will be done after MINPV is supported in the same way 
as the industry standard simulator. 

Tested on Norne. It reduces the discrepancy in well C-1H.

### CODE REVIEW

File: opm/autodiff/GeoProps.hpp [hunk 1]
```diff
@@ -186,6 +192,11 @@ namespace Opm
                                      const double* perm,
                                      Vector &hTrans);
 
+        template <class Grid>
+        void minPvFillProps_(const Grid &grid,
```
Comment (atgeirr) [inline opm/autodiff/GeoProps.hpp]: Minor nitpick (as usual): please use the standard C++ convention of attaching the `&` to the type not the variable name. It looks like taking the pointer to grid.

File: opm/autodiff/GeoProps.hpp [hunk 2]
```diff
@@ -197,6 +208,36 @@ namespace Opm
...
+        EclipseGridConstPtr eclgrid = eclState->getEclipseGrid();
+        std::vector<double> porv = eclState->getDoubleGridProperty("PORV")->getData();
+        for (int cellIdx = 0; cellIdx < numCells; ++cellIdx) {
+            const int nx = grid.cartdims[0];
+            const int ny = grid.cartdims[1];
+            const int cartesianCellIdx = global_cell[cellIdx];
+            const int cartesianCellIdxAbove = cartesianCellIdx - nx*ny;
+
+            // Average properties when the cell above exist and has porv less than the MINPV threshold
+            if ( cartesianCellIdxAbove > 0 && porv[cartesianCellIdxAbove] > 0 && porv[cartesianCellIdxAbove] < eclgrid->getMinpvValue() ) {
```
Comment (atgeirr) [inline opm/autodiff/GeoProps.hpp]: Should not this be `cartesianCellIdxAbove >= 0`?

File: opm/autodiff/GeoProps.hpp [hunk 3]
```diff
@@ -197,6 +208,36 @@ namespace Opm
...
+
+                // Volume weighted arithmetic average of NTG
+                const double cellVolume = eclgrid->getCellVolume(cartesianCellIdx);
+                const double cellAboveVolume = eclgrid->getCellVolume(cartesianCellIdxAbove);
+                double inv_totalCellVolume = 1/(cellVolume + cellAboveVolume);
+
+                ntg[cartesianCellIdx] *= vol;
```
Comment (atgeirr) [inline opm/autodiff/GeoProps.hpp]: I may be blind, but I cannot find a definition of `vol`. Should it be `cellVolume`?

Comment (atgeirr) [inline opm/autodiff/GeoProps.hpp]: What happens if more than one layer of cells are removed by MINPV? I think this code will only include the single layer above and not all layers above that are removed.

**Comment by atgeirr** [comment]
Thanks for the patch! This should definitely go in, the only serious question I have is the one about multiple layers. It would improve even in that case though, so if it is hard to fix I am open to merging without that.

**Comment by totto82** [comment]
The code should now handle multiple layers of MINPV merged cells. (There are around 20 cases of this in Norne). Testing on Norne in progress. I will report back when the testing is complete.

File: opm/autodiff/GeoProps.hpp (hunk 1 above)
Comment (totto82) [inline opm/autodiff/GeoProps.hpp]: I just followed the convention in the rest of the file.

Comment (atgeirr) [inline opm/autodiff/GeoProps.hpp]: I see, did not notice. If I find a tool that automatically prettifies all code in a pleasing manner I'll run it on the entire code base... Until then, imitating the style in the file you are editing is the best approach.

**Comment by atgeirr** [comment]
Testing has uncovered minor problems with this PR, it will therefore be postponed until after the upcoming release.

**Comment by totto82** [comment]
> Testing has uncovered minor problems with this PR, it will therefore be postponed until after the upcoming release.

With the current Norne data file, ran with the parameters in run_norne_opm.param, the simulation runs through with results that are closer to Eclipse for the effected wells. (and same for the rest)

**Comment by atgeirr** [comment]
So we should therefore merge it as-is?

**Comment by totto82** [comment]
I don't think the instability or (sensitivity to input parameters) in convergence we experience is directly related to this PR. If the current parameters in opm-data is regarded as "the default" and the main prerequisite for Release 2015.04 is ability to run norne. Then I will say that this PR should be merged as-is.

**Comment by bska** [comment]
```
/home/bska/work/opm/src/opm/opm-autodiff/opm/autodiff/GeoProps.hpp:222: error: ‘const class Dune::CpGrid’ has no member named ‘cartdims’
/home/bska/work/opm/src/opm/opm-autodiff/opm/autodiff/GeoProps.hpp:223: error: ‘const class Dune::CpGrid’ has no member named ‘cartdims’
```

**Comment by atgeirr** [comment]
Can confirm that I get the same error: a bit surprised, because since you have run tests with this it must have compiled!?!

Anyway: I'll try to fix this as quickly as I can.

**Comment by bska** [comment]
> a bit surprised, because since you have run tests with this it must have compiled

This happens "only" when opm-autodiff depends on dune-cornerpoint.

**Comment by atgeirr** [comment]
> This happens "only" when opm-autodiff depends on dune-cornerpoint

Of course, thanks! I guess I should have been a little bit more diligent in the reviewing then, I tend to forget that it is optional!

**Pull request merged**